modules/standard/List.chpl

```diff
@@ -1318,20 +1324,152 @@ module List {
+      }
+
+      return _getRef(i);
+    }
+
+    /*
+      Return a borrow of the element at a given index in this list. This
+      method can only be called when this list's element type is a class
+      type.
+
+      :arg i: The index of the element to borrow.
+      :type i: `int`
+
+      :return: A borrow of an element from this list.
+    */
+    proc const getBorrowed(i: int) where isClass(eltType) {
+      _enter(); defer _leave();
+
+      if boundsChecking && !_withinBounds(i) {
+        const msg = "Invalid list index: " + i:string;
+        boundsCheckHalt(msg);
+      }
+
+      ref slot = _getRef(i);
+
+      return slot.borrow();
+    }
+
+    /*
+      Sets the element at a given index in this list. This method returns
+      `false` if the index is out of bounds.
+
+      :arg i: The index of the element to set
+      :type i: int
+
+      :arg x: The value to set at index `i`
+
+      :return: `true` if `i` is a valid index that has been set to `x`,
+               and `false` otherwise.
+      :rtype: bool
+    */
+    proc ref set(i: int, pragma "no auto destroy" in x: eltType): bool {
+      _enter(); defer _leave();
+
+      if !_withinBounds(i) {
+        _destroy(x);
+        return false;
+      }
+
+      ref src = x;
+      ref dst = _getRef(i);
+      _move(src, dst);
+
+      return true;
+    }
+
+    /*
+      Update a value in this list in a parallel safe manner via a worker
+      object.
+
+      The worker object passed to the `update()` method must
+      define a `this()` method that takes two arguments: an integer index,
+      and a second argument of this list's `valType`. The worker object's
+      `this()` method must return some sort of value. Workers that do not
+      need to return anything may return `none`.
+
+      If the worker's `this()` method throws, the thrown error will be
+      propagated out of `update()`.
+
+      :arg i: The index to update
+      :type i: `int`
+
+      :arg worker: A class or record used to update the value at `i`
+      :return: What the worker returns
+    */
+    proc update(i: int, worker) throws {
+      _enter(); defer _leave();
+
+      if boundsChecking && !_withinBounds(i) {
+        const msg = "Invalid list index: " + i:string;
+        boundsCheckHalt(msg);
+      }
+
+      ref slot = _getRef(i);
+
+      // Print a prettier error message if arguments fail resolve, to avoid
+      // pointing into module code.
+      import Reflection;
+      if !Reflection.canResolveMethod(worker, "this", i, slot) then
+        compilerError('`list.update()` failed to resolve method ' +
+                      worker.type:string + '.this() for arguments (' +
+                      i.type:string + ', ' + slot.type:string + ')');
+
+      return worker(i, slot);
+    }
+
+    pragma "no doc"
+    inline proc _errorForParSafeIndexing() {
+      if parSafe then
+        compilerError('Cannot index a list initialized with ' +
+                      '`parSafe=true`', 2);
```

Comment:
Should this and the error for first/last recommend using update or set or something? Additionally, since these used to work, do we need to deprecate it for now (i.e. make these a warning) rather than outright ban it?

Reply:
Do deprecation warnings for `this`, `first`, and `last`, sound good?

Reply:
yes

Reply:
I've gone with deprecation warnings and suggested using `set()` or `update()`.